File: stock_alerts.php
# stock_alerts.php
#!/usr/bin/env php
<?php

define('CONFIG_FILE', 'alerts.json');
define('API_KEY', 'demo');
define('BASE_URL', 'https://www.alphavantage.co/query');

class Alert {
    public $id;
    public $symbol;
    public $threshold;
    public $above;
    public $triggered;

    function __construct($symbol, $threshold, $above = true, $id = null) {
        $this->id = $id ?: substr(bin2hex(random_bytes(4)), 0, 8);
        $this->symbol = strtoupper($symbol);
        $this->threshold = (float)$threshold;
        $this->above = $above;
        $this->triggered = false;
    }

    function toArray() {
        return [
            'id' => $this->id,
            'symbol' => $this->symbol,
            'threshold' => $this->threshold,
            'above' => $this->above,
            'triggered' => $this->triggered
        ];
    }

    static function fromArray($data) {
        $a = new self($data['symbol'], $data['threshold'], $data['above'], $data['id']);
        $a->triggered = $data['triggered'] ?? false;
        return $a;
    }

    function check($price) {
        $triggered = $this->above ? $price > $this->threshold : $price < $this->threshold;
        if ($triggered && !$this->triggered) {
            $this->triggered = true;
            return true;
        }
        return false;
    }
}

class App {
    private $alerts = [];

    function __construct() {
        $this->load();
    }

    function load() {
        if (file_exists(CONFIG_FILE)) {
            $data = json_decode(file_get_contents(CONFIG_FILE), true);
            foreach ($data as $item) {
                $this->alerts[] = Alert::fromArray($item);
            }
        }
    }

    function save() {
        $data = array_map(function($a) { return $a->toArray(); }, $this->alerts);
        file_put_contents(CONFIG_FILE, json_encode($data, JSON_PRETTY_PRINT));
    }

    function add($symbol, $threshold, $above = true) {
        $this->alerts = array_filter($this->alerts, function($a) use ($symbol) {
            return $a->symbol != strtoupper($symbol);
        });
        $alert = new Alert($symbol, $threshold, $above);
        $this->alerts[] = $alert;
        $this->save();
        echo "✅ Added alert: $symbol " . ($above ? '>' : '<') . " " . number_format($threshold, 2) . "\n";
    }

    function remove($symbol) {
        $this->alerts = array_filter($this->alerts, function($a) use ($symbol) {
            return $a->symbol != strtoupper($symbol);
        });
        $this->save();
        echo "✅ Removed alert for " . strtoupper($symbol) . "\n";
    }

    function list() {
        if (empty($this->alerts)) {
            echo "No alerts.\n";
            return;
        }
        echo "\n📋 Current alerts:\n";
        foreach ($this->alerts as $a) {
            $status = $a->triggered ? '🔔' : '⏳';
            echo "  {$a->symbol}: " . ($a->above ? '>' : '<') . " " . number_format($a->threshold, 2) . " $status\n";
        }
    }

    function fetchPrice($symbol) {
        $url = BASE_URL . "?function=GLOBAL_QUOTE&symbol=$symbol&apikey=" . API_KEY;
        $resp = file_get_contents($url);
        if ($resp === false) return null;
        $data = json_decode($resp, true);
        if (!isset($data['Global Quote'])) return null;
        return (float)$data['Global Quote']['05. price'];
    }

    function checkAlerts() {
        foreach ($this->alerts as $alert) {
            $price = $this->fetchPrice($alert->symbol);
            if ($price === null) continue;
            if ($alert->check($price)) {
                $direction = $alert->above ? 'above' : 'below';
                echo "\n🔔 ALERT: {$alert->symbol} price is " . number_format($price, 2) . " ($direction " . number_format($alert->threshold, 2) . ")!\n";
                $this->save();
            }
        }
    }

    function watch($interval = 60) {
        echo "📈 Monitoring alerts every {$interval}s (press Ctrl+C to stop)\n";
        while (true) {
            $this->checkAlerts();
            sleep($interval);
        }
    }
}

if ($argc < 2) {
    die("Usage: php stock_alerts.php <command> [options]\n");
}
$app = new App();
$cmd = $argv[1];

switch ($cmd) {
    case 'add':
        if ($argc < 4) die("add <symbol> <threshold> [--above] [--below]\n");
        $symbol = $argv[2];
        $threshold = (float)$argv[3];
        $above = true;
        for ($i=4; $i<$argc; $i++) {
            if ($argv[$i] == '--below') $above = false;
        }
        $app->add($symbol, $threshold, $above);
        break;
    case 'remove':
        if ($argc < 3) die("remove <symbol>\n");
        $app->remove($argv[2]);
        break;
    case 'list':
        $app->list();
        break;
    case 'watch':
        $interval = 60;
        for ($i=2; $i<$argc; $i++) {
            if ($argv[$i] == '--interval' && isset($argv[$i+1])) {
                $interval = (int)$argv[++$i];
            }
        }
        $app->watch($interval);
        break;
    default:
        echo "Unknown command.\n";
}
?>

File: stock_alerts.rb
# stock_alerts.rb
#!/usr/bin/env ruby
require 'json'
require 'net/http'
require 'uri'
require 'time'
require 'securerandom'
require 'optparse'

CONFIG_FILE = 'alerts.json'
API_KEY = 'demo'
BASE_URL = 'https://www.alphavantage.co/query'

class Alert
  attr_accessor :id, :symbol, :threshold, :above, :triggered

  def initialize(symbol, threshold, above = true, id = nil)
    @id = id || SecureRandom.hex(4)
    @symbol = symbol.upcase
    @threshold = threshold
    @above = above
    @triggered = false
  end

  def check(price)
    triggered = @above ? price > @threshold : price < @threshold
    if triggered && !@triggered
      @triggered = true
      return true
    end
    false
  end

  def to_hash
    { id: @id, symbol: @symbol, threshold: @threshold, above: @above, triggered: @triggered }
  end

  def self.from_hash(h)
    alert = new(h['symbol'], h['threshold'], h['above'], h['id'])
    alert.triggered = h['triggered'] || false
    alert
  end
end

class App
  attr_reader :alerts

  def initialize
    @alerts = []
    load
  end

  def load
    if File.exist?(CONFIG_FILE)
      data = JSON.parse(File.read(CONFIG_FILE))
      @alerts = data.map { |h| Alert.from_hash(h) }
    end
  end

  def save
    File.write(CONFIG_FILE, JSON.pretty_generate(@alerts.map(&:to_hash)))
  end

  def add(symbol, threshold, above = true)
    @alerts.reject! { |a| a.symbol == symbol.upcase }
    alert = Alert.new(symbol, threshold, above)
    @alerts << alert
    save
    puts "✅ Added alert: #{symbol} #{above ? '>' : '<'} #{'%.2f' % threshold}"
  end

  def remove(symbol)
    @alerts.reject! { |a| a.symbol == symbol.upcase }
    save
    puts "✅ Removed alert for #{symbol.upcase}"
  end

  def list
    if @alerts.empty?
      puts "No alerts."
      return
    end
    puts "\n📋 Current alerts:"
    @alerts.each do |a|
      status = a.triggered ? '🔔' : '⏳'
      puts "  #{a.symbol}: #{a.above ? '>' : '<'} #{'%.2f' % a.threshold} #{status}"
    end
  end

  def fetch_price(symbol)
    uri = URI("#{BASE_URL}?function=GLOBAL_QUOTE&symbol=#{symbol}&apikey=#{API_KEY}")
    response = Net::HTTP.get(uri)
    data = JSON.parse(response)
    quote = data['Global Quote']
    return nil unless quote
    quote['05. price'].to_f
  rescue => e
    warn "Error fetching #{symbol}: #{e.message}"
    nil
  end

  def check_alerts
    @alerts.each do |alert|
      price = fetch_price(alert.symbol)
      next unless price
      if alert.check(price)
        direction = alert.above ? 'above' : 'below'
        puts "\n🔔 ALERT: #{alert.symbol} price is #{'%.2f' % price} (#{direction} #{'%.2f' % alert.threshold})!"
        save
      end
    end
  end

  def watch(interval = 60)
    puts "📈 Monitoring alerts every #{interval}s (press Ctrl+C to stop)"
    loop do
      check_alerts
      sleep(interval)
    end
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: stock_alerts.rb [command] [options]"
  opts.on("add SYMBOL THRESHOLD", "Add alert") { |v| options[:add] = v }
  opts.on("remove SYMBOL", "Remove alert") { |v| options[:remove] = v }
  opts.on("list", "List alerts") { options[:list] = true }
  opts.on("watch", "Start monitoring") { options[:watch] = true }
  opts.on("--above", "Alert above threshold") { options[:above] = true }
  opts.on("--below", "Alert below threshold") { options[:below] = true }
  opts.on("--interval N", Integer, "Watch interval") { |v| options[:interval] = v }
end.parse!

app = App.new

if options[:add]
  symbol, threshold = options[:add].split
  above = !options[:below]
  app.add(symbol, threshold.to_f, above)
elsif options[:remove]
  app.remove(options[:remove])
elsif options[:list]
  app.list
elsif options[:watch]
  interval = options[:interval] || 60
  app.watch(interval)
else
  puts "Unknown command. Use add, remove, list, watch."
end
